cpp/src/arrow/compute/kernels/filter.cc

```diff
@@ -163,5 +164,79 @@ Status Filter(FunctionContext* ctx, const RecordBatch& batch, const Array& filte
   return Status::OK();
 }
 
+Status Filter(FunctionContext* ctx, const ChunkedArray& values, const Array& filter,
+              std::shared_ptr<ChunkedArray>* out) {
+  auto num_chunks = values.num_chunks();
+  std::vector<std::shared_ptr<arrow::Array>> new_chunks(num_chunks);
+  std::shared_ptr<arrow::Array> current_chunk;
+  int offset = 0;
+  int len;
+
+  for (int i = 0; i < num_chunks; i++) {
+    current_chunk = values.chunk(i);
+    len = current_chunk->length();
+    RETURN_NOT_OK(Filter(ctx, *current_chunk, *filter.Slice(offset, len), &new_chunks[i]));
+    offset += len;
+  }
+
+  *out = std::make_shared<arrow::ChunkedArray>(std::move(new_chunks));
+  return Status::OK();
+}
+
+Status Filter(FunctionContext* ctx, const ChunkedArray& values, const ChunkedArray& filter,
+              std::shared_ptr<ChunkedArray>* out) {
+  auto num_chunks = values.num_chunks();
+  std::vector<std::shared_ptr<arrow::Array>> new_chunks(num_chunks);
+  std::shared_ptr<arrow::Array> current_chunk;
+  std::shared_ptr<arrow::ChunkedArray> current_chunked_filter;
+  std::shared_ptr<arrow::Array> current_filter;
+  int offset = 0;
+  int len;
+
+  for (int i = 0; i < num_chunks; i++) {
+    current_chunk = values.chunk(i);
+    len = current_chunk->length();
+    current_chunked_filter = filter.Slice(offset, len);
+    if (current_chunked_filter->num_chunks() == 1) {
+      current_filter = current_chunked_filter->chunk(0);
+    } else {
+      // Concatenate the chunks of the filter so we have an Array
+      RETURN_NOT_OK(Concatenate(current_chunked_filter->chunks(),
+                                        default_memory_pool(), &current_filter));
+    }
+    RETURN_NOT_OK(Filter(ctx, *current_chunk, *current_filter, &new_chunks[i]));
+    offset += len;
+  }
+
+  *out = std::make_shared<arrow::ChunkedArray>(std::move(new_chunks));
+  return Status::OK();
+}
+
+Status Filter(FunctionContext* ctx, const Table& table, const Array& filter,
+              std::shared_ptr<Table>* out) {
+  auto ncols = table.num_columns();
+
+  std::vector<std::shared_ptr<arrow::ChunkedArray>> columns(ncols);
+
+  for (int j = 0; j < ncols; j++) {
+    RETURN_NOT_OK(Filter(ctx, *table.column(j), filter, &columns[j]));
+  }
+  *out = Table::Make(table.schema(), columns);
+  return Status::OK();
+}
+
+Status Filter(FunctionContext* ctx, const Table& table, const ChunkedArray& filter,
+              std::shared_ptr<Table>* out) {
+  auto ncols = table.num_columns();
+
+  std::vector<std::shared_ptr<arrow::ChunkedArray>> columns(ncols);
```

Comment:
Another missed (deferred) optimization would be to parallelize the filtering across the columns rather than iterating in serial. 

I'll take a look at the one you reference and see if I can translate it for tables; otherwise I'll make followup Jiras.